Unable to generate manifest for files from 1000G snapshot

Reproduction:

- Attempt to create a manifest on `anvildev` that includes files from the 1000G snapshot:

```
$ curl -X PUT 'https://service.anvil.gi.ucsc.edu/fetch/manifest/files?catalog=anvil&format=compact'
{"Status":301,"Location":"https://service.anvil.gi.ucsc.edu/fetch/manifest/files/lMQQurNW7O3hUJitVpRZXv8vWAEAAQ==","Retry-After":1}

$ curl 'https://service.anvil.gi.ucsc.edu/fetch/manifest/files/lMQQurNW7O3hUJitVpRZXv8vWAEAAQ=='
{"Code":"ChaliceViewError","Message":"('Failed to generate manifest', 'FAILED', None)"}
```


**CloudWatch Logs Insights**    
log-group-names: /aws/lambda/azul-service-anvildev-manifest    

---
| @timestamp | @message |
| --- | --- |
| 2025-12-13 03:27:46.799 | INIT_START Runtime Version: python:3.13.v76 Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:1a6363019b274fc28ffbbec073e5ebf1a872c10c52269844a57c62e74063a49b |
| 2025-12-13 03:27:49.555 | START RequestId: 4ac8de91-0498-4346-ae97-5c2831d2795a Version: $LATEST |
| 2025-12-13 03:27:49.571 | [INFO] 2025-12-13T03:27:49.571Z 4ac8de91-0498-4346-ae97-5c2831d2795a botocore.credentials Found credentials in environment variables. |
| 2025-12-13 03:27:49.667 | [INFO] 2025-12-13T03:27:49.666Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.deployment Allocated new Boto3 client for &#x27;sts&#x27; with ID 139965883160896 |
| 2025-12-13 03:27:49.786 | [INFO] 2025-12-13T03:27:49.786Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.deployment Allocated new Boto3 client for &#x27;s3&#x27; with ID 139965794232912 |
| 2025-12-13 03:27:49.791 | [INFO] 2025-12-13T03:27:49.791Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.HeadObject: Making HEAD request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-dev-storage-anvildev.us-east-1&#x2f;manifests&#x2f;690f2a1d-5232-559a-9427-11d698f2e34c.9e52f9c4-c112-5342-a470-ab60ee76298a.tsv |
| 2025-12-13 03:27:49.791 | [INFO] 2025-12-13T03:27:49.791Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.HeadObject: … without a request body |
| 2025-12-13 03:27:49.848 | [INFO] 2025-12-13T03:27:49.848Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.HeadObject: Got 404 response |
| 2025-12-13 03:27:49.848 | [INFO] 2025-12-13T03:27:49.848Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.HeadObject: … without a response body |
| 2025-12-13 03:27:49.849 | [INFO] 2025-12-13T03:27:49.848Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.service.manifest_service Cached manifest not found: ManifestKey(catalog=&#x27;anvil&#x27;, format=&lt;ManifestFormat.compact: &#x27;compact&#x27;&gt;, manifest_hash=UUID(&#x27;690f2a1d-5232-559a-9427-11d698f2e34c&#x27;), source_hash=UUID(&#x27;9e52f9c4-c112-5342-a470-ab60ee76298a&#x27;)) |
| 2025-12-13 03:27:49.851 | [INFO] 2025-12-13T03:27:49.851Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.CreateMultipartUpload: Making POST request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-dev-storage-anvildev.us-east-1&#x2f;manifests&#x2f;690f2a1d-5232-559a-9427-11d698f2e34c.9e52f9c4-c112-5342-a470-ab60ee76298a.tsv?uploads |
| 2025-12-13 03:27:49.851 | [INFO] 2025-12-13T03:27:49.851Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.CreateMultipartUpload: … without a request body |
| 2025-12-13 03:27:49.882 | [INFO] 2025-12-13T03:27:49.882Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.CreateMultipartUpload: Got 200 response |
| 2025-12-13 03:27:49.882 | [INFO] 2025-12-13T03:27:49.882Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.boto3 s3.CreateMultipartUpload: … with a response body of length 515 being b&#x27;&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;\n&lt;InitiateMultipartUploadResult xmlns=&quot;http:&#x2f;&#x2f;s3.amazonaws.com&#x2f;doc&#x2f;2006-03-01&#x2f;&quot;&gt;&lt;Bucket&gt;edu-ucsc-gi-platform-anvil-dev-storage-anvildev.us-east-1&lt;&#x2f;Bucket&gt;&lt;Key&gt;manifests&#x2f;690f2a1d-5232-559a-9427-11d698f2e34c.9e52f9c4-c112-5342-a470-ab60ee76298a.tsv&lt;&#x2f;Key&gt;&lt;UploadId&gt;cvOFr...-&lt;&#x2f;UploadId&gt;&lt;&#x2f;InitiateMultipartUploadResult&gt;&#x27; |
| 2025-12-13 03:27:49.908 | [DEBUG] 2025-12-13T03:27:49.907Z 4ac8de91-0498-4346-ae97-5c2831d2795a azul.es Creating ES client [vpc-azul-index-anvildev-jubbechxp4wxzfj2kf2hiewjgi.us-east-1.es.amazonaws.com:443] |
| 2025-12-13 03:27:49.918 | [INFO] 2025-12-13T03:27:49.918Z 4ac8de91-0498-4346-ae97-5c2831d2795a botocore.credentials Found credentials in environment variables. |
| 2025-12-13 03:27:49.921 | [INFO] 2025-12-13T03:27:49.921Z 4ac8de91-0498-4346-ae97-5c2831d2795a opensearch Making POST request to https:&#x2f;&#x2f;vpc-azul-index-anvildev-jubbechxp4wxzfj2kf2hiewjgi.us-east-1.es.amazonaws.com:443&#x2f;azul_v2_anvildev_anvil_files_aggregate&#x2f;_search |
| 2025-12-13 03:27:49.921 | [INFO] 2025-12-13T03:27:49.921Z 4ac8de91-0498-4346-ae97-5c2831d2795a opensearch … with a request body of length 8731 starting in b&#x27;{&quot;query&quot;:{&quot;bool&quot;:{&quot;must&quot;:[{&quot;constant_score&quot;:{&quot;filter&quot;:{&quot;terms&quot;:{&quot;sources.id.keyword&quot;:[&quot;001afc86-da4c-4739-85be-26ca98d2693f&quot;,&quot;006f69b6-7f42-4eba-81df-86805c22dd3d&quot;,&quot;01b1ac88-8144-428e-a84c-88e9ed8ec6a2&quot;,&quot;02b9f4e0-3bee-4c23-87d4-485ddb414a87&quot;,&quot;03b58060-2316-42fc-811f-0e5895399e96&quot;,&quot;0522bcb4-e8f9-4c0b-936b-b2d729cbcac7&quot;,&quot;0559efce-d188-4a59-9f6f-2a6258ca7e79&quot;,&quot;057262af-09d8-4975-8647-2a68a5d33629&quot;,&quot;05ed25dc-408d-4045-998d-6166fc2c3293&quot;,&quot;06eca349-0e2c-4785-a93a-1eeb46a50220&quot;,&quot;07bd9e9a-f3d5-4ce7-a7b5-a82cf5ede5b9&quot;,&quot;0a4bd5cf-f007-49b2-98e7-f328490ede47&quot;,&quot;0a711aa4-24d1-46e2-9d4b-5b02887311f9&quot;,&quot;0ab78bd0-2e02-4536-b3c9-bdd7a57d7988&quot;,&quot;0c86170e-312d-4b39-a0a4-2a2bfaa24c7a&quot;,&quot;0e5501d1-39ca-44bd-b2a8-127841eb19ba&quot;,&quot;104f2254-2af9-4699-9573-00d2da718430&quot;,&quot;139ecf31-b6bb-419a-97c8-1245029631bb&quot;,&quot;13c6bf33-41a3-4330-a3f7-1e69f675b01e&quot;,&quot;15242c5c-0c11-441a-8b63-4cafcebe892f&quot;,&quot;155a019c-ca47-432d-939b-6e51d5bec730&quot;,&quot;1568e68d-3032-4571-a066-dac6ccec9716&quot;,&quot;170988f8-1dd1-4f39-b72d-645dabf607b7&quot;,&quot;1719e522-0e4d-4149-8952-b723dc5d304a&quot;...&#x27; |
| 2025-12-13 03:27:50.048 | [INFO] 2025-12-13T03:27:50.048Z 4ac8de91-0498-4346-ae97-5c2831d2795a opensearch Got 200 response after 0.125s from POST to https:&#x2f;&#x2f;vpc-azul-index-anvildev-jubbechxp4wxzfj2kf2hiewjgi.us-east-1.es.amazonaws.com:443&#x2f;azul_v2_anvildev_anvil_files_aggregate&#x2f;_search |
| 2025-12-13 03:27:50.048 | [INFO] 2025-12-13T03:27:50.048Z 4ac8de91-0498-4346-ae97-5c2831d2795a opensearch … with a response body of length 1552389 starting in &#x27;{&quot;took&quot;:77,&quot;timed_out&quot;:false,&quot;_shards&quot;:{&quot;total&quot;:1,&quot;successful&quot;:1,&quot;skipped&quot;:0,&quot;failed&quot;:0},&quot;hits&quot;:{&quot;total&quot;:{&quot;value&quot;:26066,&quot;relation&quot;:&quot;eq&quot;},&quot;max_score&quot;:null,&quot;hits&quot;:[{&quot;_index&quot;:&quot;azul_v2_anvildev_anvil_files_aggregate&quot;,&quot;_id&quot;:&quot;0001b016-e73e-4f2e-a64d-b51555a69af3&quot;,&quot;_score&quot;:null,&quot;_source&quot;:{&quot;sources&quot;:[{&quot;id&quot;:&quot;790795c4-49b1-4ac8-a060-207b92ea08c5&quot;,&quot;spec&quot;:&quot;tdr:bigquery:gcp:datarepo-dev-e53e74aa:ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732&quot;}],&quot;contents&quot;:{&quot;activities&quot;:[{&quot;assay_type&quot;:[&quot;~null&quot;],&quot;activity_type&quot;:[&quot;Checksum&quot;,&quot;Indexing&quot;,&quot;Unknown&quot;],&quot;activity_id&quot;:[&quot;083d2d61-ecc6-1c90-23f3-661f891217de&quot;,&quot;1aa50c32-38a1-2180-b468-78a4930c2491&quot;,&quot;85e23b00-1f04-2d25-43e7-9550ebda40f9&quot;,&quot;d11a5e55-8f5a-4129-6a2d-9dca04f3c023&quot;],&quot;reference_assembly&quot;:[&quot;~null&quot;],&quot;document_id&quot;:[&quot;1d321318-c6f5-41d7-a770-3e0e9c231dc9&quot;,&quot;3d6ede20-c052-43dd-860b-be432038864f&quot;,&quot;b0b87e8f-5a36-4df7-976f-2df753abc41e&quot;,&quot;c847cccf-627d-404d-8e87-2e907f44ccc6&quot;],&quot;data_modality&quot;:[&quot;~null&quot;],&quot;source_datarepo_row_ids&quot;:[&quot;file_inventory:56afa2a2-e792-44dc-b6a5-0b82bfc82edd&quot;,&quot;fil…&#x27; |
| 2025-12-13 03:27:50.086 | [ERROR] AssertionError: &lt;class &#x27;NoneType&#x27;&gt; Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1707, in __call__     return self._original_func(event.to_dict(), event.context)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 1351, in generate_manifest     return app.manifest_controller.get_manifest(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_controller.py&quot;, line 96, in get_manifest     result = self.service.get_manifest(format=manifest_key.format,   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_service.py&quot;, line 641, in get_manifest     return self._generate_manifest(generator, manifest_key, partition)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_service.py&quot;, line 650, in _generate_manifest     partition = generator.write(manifest_key, partition)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_service.py&quot;, line 1327, in write     partition = self.write_page_to(partition, output=text_buffer)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_service.py&quot;, line 1694, in write_page_to     file[&#x27;file_mirror_uri&#x27;] = self._azul_mirror_uri(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;manifest_service.py&quot;, line 1159, in _azul_mirror_uri     file = self.metadata_plugin.file_class.from_index(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;plugins&#x2f;metadata&#x2f;anvil&#x2f;__init__.py&quot;, line 551, in from_index     md5=json_str(hit[&#x27;file_md5sum&#x27;]))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;types.py&quot;, line 165, in json_str     return any_str(v)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;types.py&quot;, line 169, in any_str     assert isinstance(v, str), type(v) |
| 2025-12-13 03:27:50.089 | END RequestId: 4ac8de91-0498-4346-ae97-5c2831d2795a |
| 2025-12-13 03:27:50.090 | REPORT RequestId: 4ac8de91-0498-4346-ae97-5c2831d2795a Duration: 533.82 ms Billed Duration: 3286 ms Memory Size: 2048 MB Max Memory Used: 169 MB Init Duration: 2751.33 ms |
---


```
[ERROR] AssertionError: <class 'NoneType'>
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 194, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1761, in __call__
    return self.handler(event_obj)
  File "/var/task/chalice/app.py", line 1707, in __call__
    return self._original_func(event.to_dict(), event.context)
  File "/var/task/app.py", line 1351, in generate_manifest
    return app.manifest_controller.get_manifest(event)
  File "/var/task/azul/service/manifest_controller.py", line 96, in get_manifest
    result = self.service.get_manifest(format=manifest_key.format,
  File "/var/task/azul/service/manifest_service.py", line 641, in get_manifest
    return self._generate_manifest(generator, manifest_key, partition)
  File "/var/task/azul/service/manifest_service.py", line 650, in _generate_manifest
    partition = generator.write(manifest_key, partition)
  File "/var/task/azul/service/manifest_service.py", line 1327, in write
    partition = self.write_page_to(partition, output=text_buffer)
  File "/var/task/azul/service/manifest_service.py", line 1694, in write_page_to
    file['file_mirror_uri'] = self._azul_mirror_uri(file)
  File "/var/task/azul/service/manifest_service.py", line 1159, in _azul_mirror_uri
    file = self.metadata_plugin.file_class.from_index(file)
  File "/var/task/azul/plugins/metadata/anvil/__init__.py", line 551, in from_index
    md5=json_str(hit['file_md5sum']))
  File "/var/task/azul/types.py", line 165, in json_str
    return any_str(v)
  File "/var/task/azul/types.py", line 169, in any_str
    assert isinstance(v, str), type(v)
```

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -38,6 +38,7 @@ from azul.functions import (
 from azul.indexer import (
     SourceConfig,
     SourceRef,
+    SourceSpec,
 )
 from azul.indexer.mirror_file_service import (
     FilePart,
@@ -134,6 +135,23 @@ class BaseMirrorService:
     @cache
     def _repository_plugin(self, catalog: CatalogName) -> RepositoryPlugin:
         return RepositoryPlugin.load(catalog).create(catalog)
+
+    def may_mirror_files_from_source(self,
+                                     catalog: CatalogName,
+                                     source_spec: SourceSpec,
+                                     ) -> bool:
+        """
+        Test whether it makes sense to request the mirroring of files from the
+        given source. If this method returns True, files from the source may or
+        may not be mirrored. If this method returns False, the service will
+        definitely refuse to mirror all files from the source.
+        """
+        if self.may_mirror(catalog):
+            plugin = self._repository_plugin(catalog)
+            source_config = plugin.sources[source_spec]
+            return source_config.mirror
+        else:
+            return False
 
     @classmethod
     def may_mirror(cls, catalog: CatalogName, file_size: int = 0) -> bool:

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -131,6 +131,10 @@ class BaseMirrorService:
     def _queues(self) -> Queues:
         return Queues()
 
+    @cache
+    def _repository_plugin(self, catalog: CatalogName) -> RepositoryPlugin:
+        return RepositoryPlugin.load(catalog).create(catalog)
+
     @classmethod
     def may_mirror(cls, catalog: CatalogName, file_size: int = 0) -> bool:
         """
@@ -198,10 +202,6 @@ class MirrorService(BaseMirrorService):
     def _file_service(self, catalog: CatalogName) -> MirrorFileService:
         return MirrorFileService(catalog=catalog,
                                  schema_url_func=self._schema_url_func)
-
-    @cache
-    def _repository_plugin(self, catalog: CatalogName) -> RepositoryPlugin:
-        return RepositoryPlugin.load(catalog).create(catalog)
 
     @cached_property
     def _source_service(self) -> SourceService:

--- a/src/azul/service/manifest_service.py
+++ b/src/azul/service/manifest_service.py
@@ -98,6 +98,8 @@ from azul.deployment import (
 )
 from azul.indexer import (
     Prefix,
+    SourceRef,
+    SourceSpec,
 )
 from azul.indexer.document import (
     DocumentType,
@@ -821,6 +823,10 @@ class ManifestGenerator(metaclass=ABCMeta):
         return self.service.metadata_plugin(self.catalog)
 
     @cached_property
+    def mirror_service(self) -> BaseMirrorService:
+        return BaseMirrorService()
+
+    @cached_property
     def mirror_file_service(self) -> BaseMirrorFileService:
         return BaseMirrorFileService(catalog=self.catalog)
 
@@ -1054,7 +1060,12 @@ class ManifestGenerator(metaclass=ABCMeta):
         if self.included_fields is None:
             document_slice = DocumentSlice()
         else:
-            document_slice = DocumentSlice(includes=list(map(dotted, self.included_fields)))
+            includes = list(map(dotted, self.included_fields))
+            # The complete set of source fields is needed to polymorphically
+            # deserialize sources from the index. The sources are used to
+            # help determine which files might be mirrored.
+            includes.append('sources.*')
+            document_slice = DocumentSlice(includes=includes)
         pipeline = self.service.create_chain(catalog=self.catalog,
                                              entity_type=self.entity_type,
                                              filters=self.filters,
@@ -1154,10 +1165,13 @@ class ManifestGenerator(metaclass=ABCMeta):
                                           fetch=False,
                                           **args))
 
-    def _azul_mirror_uri(self, file: JSON) -> str | None:
-        file = self.metadata_plugin.file_class.from_index(file)
-        if BaseMirrorService.may_mirror(self.catalog, file.size):
-            return self.mirror_file_service.mirror_uri(file)
+    def _azul_mirror_uri(self, file: JSON, source: SourceSpec) -> str | None:
+        if self.mirror_service.may_mirror_files_from_source(self.catalog, source):
+            file = self.metadata_plugin.file_class.from_index(file)
+            if self.mirror_service.may_mirror(self.catalog, file.size):
+                return self.mirror_file_service.mirror_uri(file)
+            else:
+                return None
         else:
             return None
 
@@ -1677,6 +1691,7 @@ class CompactManifestGenerator(PagedManifestGenerator):
                 doc = self._hit_to_doc(hit)
                 assert isinstance(doc, dict)
                 contents = doc['contents']
+                source = SourceRef.from_json(one(doc['sources'])).spec
                 if len(project_short_names) < 2 and 'projects' in contents:
                     project = one(cast(JSONs, contents['projects']))
                     short_names = project['project_short_name']
@@ -1690,7 +1705,7 @@ class CompactManifestGenerator(PagedManifestGenerator):
                         if 'file_url' in column_mapping:
                             file['file_url'] = self._azul_file_url(file)
                         if 'file_mirror_uri' in column_mapping:
-                            file['file_mirror_uri'] = self._azul_mirror_uri(file)
+                            file['file_mirror_uri'] = self._azul_mirror_uri(file, source)
                         entities = [file]
                     self._extract_fields(field_path=field_path,
                                          entities=entities,
@@ -1706,7 +1721,7 @@ class CompactManifestGenerator(PagedManifestGenerator):
                                 if 'file_url' in column_mapping:
                                     file['file_url'] = self._azul_file_url(file)
                                 if 'file_mirror_uri' in column_mapping:
-                                    file['file_mirror_uri'] = self._azul_mirror_uri(file)
+                                    file['file_mirror_uri'] = self._azul_mirror_uri(file, source)
                                 self._extract_fields(field_path=field_path,
                                                      entities=[file],
                                                      column_mapping=column_mapping,

--- a/src/azul/service/manifest_service.py
+++ b/src/azul/service/manifest_service.py
@@ -1065,8 +1065,7 @@ class ManifestGenerator(metaclass=ABCMeta):
     def _hit_to_doc(self, hit: Hit) -> MutableJSON:
         return self.service.translate_fields(self.catalog,
                                              hit.to_dict(),
-                                             forward=False,
-                                             allowed_paths=self.included_fields)
+                                             forward=False)
 
     column_joiner = config.manifest_column_joiner
     padded_joiner = ' ' + column_joiner + ' '

